pkg/util/server/error.go

```diff
@@ -60,10 +63,22 @@ func ClientHTTPStatusAndError(err error) (int, error) {
 		return http.StatusBadRequest, err
 	case errors.Is(err, user.ErrNoOrgID):
 		return http.StatusBadRequest, err
+	case isRPC && (int(s.Code())/100 == 4 || int(s.Code())/100 == 5):
```

Comment:
Wouldn't the logic be more intuitive if you separate cases where you need `isRPC` to be true from those where you don't need it?
```
s, isRPC := status.FromError(err) // I also recommend grpcutil.ErrorToStatus(err)
if isRPC {
  if s.Code() == codes.DeadlineExceeded {
    return http.StatusGatewayTimeout, errors.New(ErrDeadlineExceeded)
  } else if int(s.Code())/100 == 4 || int(s.Code())/100 == 5 {
    return int(s.Code()), errors.New(s.Message())
  }
  return http.StatusInternalServerError, err
}
// at this point, it is not a gRPC error, so both status.FromError(err) or httpgrpc.HTTPResponseFromError(err) would return false
switch {
  case errors.Is(err, context.Canceled) || 
	(errors.As(err, &promErr) && errors.Is(promErr.Err, context.Canceled)):
    return StatusClientClosedRequest, errors.New(ErrClientCanceled)
  case errors.Is(err, context.DeadlineExceeded): 
    return http.StatusGatewayTimeout, errors.New(ErrDeadlineExceeded)
    case errors.As(err, &queryErr):
      return http.StatusBadRequest, err
    case errors.Is(err, logqlmodel.ErrLimit) || errors.Is(err, logqlmodel.ErrParse) || errors.Is(err, logqlmodel.ErrPipeline) || errors.Is(err, logqlmodel.ErrBlocked) || errors.Is(err, logqlmodel.ErrParseMatchers):
      return http.StatusBadRequest, err
    case errors.Is(err, user.ErrNoOrgID):
      return http.StatusBadRequest, err
    default:
      return http.StatusInternalServerError, err
}
```

Reply:
Note: all these `errors.Is()` and `errors.As()` could be expensive. Imagine that most of the time you end in the `default` case. You would, at this point, try to execute all the `errors.Is()` and `errors.As()`, and only then you would return `http.StatusInternalServerError, err`. You might think about how to improve that.

Reply:
Thanks for the feedback. It's actually worse. `error.As` and `error.Is` is used to determine the code on the [querier side](https://github.com/grafana/loki/blob/main/pkg/querier/worker/util.go#L113). The error is then packaged in the `httpgrpc.HTTPResponse`. That triggers the default case of this method on the [frontend](https://github.com/grafana/loki/blob/main/pkg/querier/queryrange/serialize.go#L69). Unless ofc the error happened within the frontend 🙈